File: content/problems/2641-cousins-in-binary-tree-ii/solutions/two-pass-level-order-replacement/Solution.java
class Solution {
    public TreeNode replaceValueInTree(TreeNode root){
        root.val=0;
        List<TreeNode> level=new ArrayList<>();
        level.add(root);
        while(!level.isEmpty()){
            int nextTotal=0;
            for(TreeNode node:level){
                if(node.left!=null)nextTotal+=node.left.val;
                if(node.right!=null)nextTotal+=node.right.val;
            }
            List<TreeNode> nextLevel=new ArrayList<>();
            for(TreeNode node:level){
                int siblingSum=(node.left==null?0:node.left.val)+(node.right==null?0:node.right.val);
                if(node.left!=null){
                    node.left.val=nextTotal-siblingSum;
                    nextLevel.add(node.left);
                }
                if(node.right!=null){
                    node.right.val=nextTotal-siblingSum;
                    nextLevel.add(node.right);
                }
            }
            level=nextLevel;
        }
        return root;
    }
}

File: content/problems/2641-cousins-in-binary-tree-ii/solutions/two-pass-level-order-replacement/Solution.kt
class Solution {
    fun replaceValueInTree(root:TreeNode?):TreeNode?{
        if(root==null)return null;
        root.`val`=0;
        var level=listOf(root);
        while(level.isNotEmpty()){
            var nextTotal=0;
            for(node in level){
                nextTotal+=node.left?.`val`?:0;
                nextTotal+=node.right?.`val`?:0
            };
            val nextLevel=mutableListOf<TreeNode>();
            for(node in level){
                val siblingSum=(node.left?.`val`?:0)+(node.right?.`val`?:0);
                node.left?.let{
                    it.`val`=nextTotal-siblingSum;
                    nextLevel.add(it)
                };
                node.right?.let{
                    it.`val`=nextTotal-siblingSum;
                    nextLevel.add(it)
                }
            };
            level=nextLevel
        };
        return root
    }
}
